### PR TITLE
Linear interpolate

### PR DESCRIPTION
**Pull request by unnonouno**
I implemented linear-interpolation function. It is about two times faster than simple implementation with math-operators, and reduces required memory.

### CODE REVIEW

File: tests/chainer_tests/functions_tests/math_tests/test_linear_interpolate.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+from chainer import functions
+from chainer import gradient_check
+from chainer import testing
+from chainer.testing import attr
+from chainer.testing import condition
+
+
+class TestLinearInterpolate(unittest.TestCase):
+
+    shape = (3, 4)
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/math_tests/test_linear_interpolate.py]: Please check `shape = ()`.

File: tests/chainer_tests/functions_tests/math_tests/test_linear_interpolate.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        self.x = numpy.random.uniform(-1, 1, self.shape).astype(numpy.float32)
+        self.y = numpy.random.uniform(-1, 1, self.shape).astype(numpy.float32)
+        self.g = numpy.random.uniform(-1, 1, self.shape).astype(numpy.float32)
+
+    def check_forward(self, p_data, x_data, y_data):
+        p = chainer.Variable(p_data)
+        x = chainer.Variable(x_data)
+        y = chainer.Variable(y_data)
+
+        z = functions.linear_interpolate(p, x, y)
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/math_tests/test_linear_interpolate.py]: Please check `z.data.dtype`.

**Comment by okuta** [comment]
LGTM except comments.

**Comment by unnonouno** [comment]
I fixed it.

File: tests/chainer_tests/functions_tests/math_tests/test_linear_interpolate.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            self.check_backward_options = {
+                'eps': 0.3, 'atol': 1e-3, 'rtol': 1e-3}
+
+    def check_forward(self, p_data, x_data, y_data):
+        p = chainer.Variable(p_data)
+        x = chainer.Variable(x_data)
+        y = chainer.Variable(y_data)
+
+        z = functions.linear_interpolate(p, x, y)
+        self.assertIs(z.data.dtype.type, self.dtype)
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/math_tests/test_linear_interpolate.py]: Please use `assertEqual`.

File: tests/chainer_tests/functions_tests/math_tests/test_linear_interpolate.py (hunk 2 above)
Comment (okuta) [inline tests/chainer_tests/functions_tests/math_tests/test_linear_interpolate.py]: Done.

**Comment by okuta** [comment]
It seems the conditions of `test_backward_cpu` test is strict.
https://travis-ci.org/pfnet/chainer/jobs/132081967

**Comment by okuta** [comment]
OK.

**Comment by philip30** [comment]
Hey I have a question. Suppose that p is learnable parameter and p could go negative in the learning process. Do you think this function still makes sense if p is negative? Thanks!

**Comment by unnonouno** [comment]
It depends on the situation, but I don't know if there is a real use case with negative value.
In actual use case, a user use `linear_interpolate` function for a gate in a function such as GRU and LSTM. In these situation the user use this function with sigmoid function or something like that.

**Pull request merged**